### DISCUSSION TITLE
Use self-insert in a function

### DISCUSSION THREAD

**Discussion by blueray453** [fish-shell/fish-shell#9838] [Q&A]
I have a keybinding like 

    bind '' _other_characters

The function looks like:

```
function _other_characters
    set --local has_selection (commandline --selection-start)
    if test -n "$has_selection"
        commandline -f kill-selection
        commandline -f self-insert
    else
        commandline -f self-insert
    end
end
```

The idea is when i press a regular key and there is a selection then it will replace the selection. Else, it will just self-insert. But when I use this setup, pressing other keys has no output.

It has been earlier done with:

```
function _other_characters
    set --local has_selection (commandline --selection-start)
    set --local first_argument $argv

    if test -n "$first_argument"
        if test -n "$has_selection"
            commandline -f kill-selection
            commandline --insert $first_argument
        else
            commandline --insert $first_argument
        end
    else
        return
    end
end
```
And the keybinding looked like:

    set --local letters 'qwertyuiopasdfghjklzxcvbnmQWERTYUIOPASDFGHJKLZXCVBNM'
    set --local numbers '1234567890'
    set --local special '!@#$%^&*_+-=\\;,./<>?:|'
    set --local pairs '\'"()[]{}'
    set --local characters (string split '' "$letters$numbers$special$pairs")

    for character in $characters
        set --local escaped (string escape -- $character)
        bind $character "_other_characters -- $escaped"
    end

In this approach, i do not think `commandline --insert` does what i understood it does. My assumption was, `commandline --insert` insert something keeping anything before or after the cursor intact. But in this particular case, string after the cursor gets changed.

**Comment by faho** [comment]
This is fundamentally #3031 - `commandline -f` functions are *queued up*, it doesn't wait for the function to be executed before returning.

So this is currently essentially impossible, sorry.

**Reply by blueray453** [reply]
I solved it by `bind '' kill-selection end-selection self-insert`.
